src/osd/scrubber/osd_scrub_sched.cc

```diff
@@ -47,298 +48,152 @@ std::ostream& ScrubQueue::gen_prefix(std::ostream& out, std::string_view fn)
-	     << ScrubJob::qu_state_text(expected_state) << " to "
-	     << ScrubJob::qu_state_text(scrub_job->state) << dendl;
-
-  } else {
-
-    // job wasn't in state 'registered' coming in
-    dout(5) << "removing pg[" << scrub_job->pgid
-	    << "] failed. State was: " << ScrubJob::qu_state_text(expected_state)
-	    << dendl;
-  }
-}
-
-void ScrubQueue::register_with_osd(
-  Scrub::ScrubJobRef scrub_job,
-  const sched_params_t& suggested)
+void ScrubQueue::remove_from_osd_queue(const Scrub::ScrubJob& scrub_job)
 {
-  qu_state_t state_at_entry = scrub_job->state.load();
-  dout(20) << fmt::format(
-		"pg[{}] state at entry: <{:.14}>", scrub_job->pgid,
-		state_at_entry)
-	   << dendl;
-
-  switch (state_at_entry) {
-    case qu_state_t::registered:
-      // just updating the schedule?
-      update_job(scrub_job, suggested, false /* keep n.b. delay */);
-      break;
-
-    case qu_state_t::not_registered:
-      // insertion under lock
-      {
-	std::unique_lock lck{jobs_lock};
-
-	if (state_at_entry != scrub_job->state) {
-	  lck.unlock();
-	  dout(5) << " scrub job state changed. Retrying." << dendl;
-	  // retry
-	  register_with_osd(scrub_job, suggested);
-	  break;
-	}
-
-	update_job(scrub_job, suggested, true /* resets not_before */);
-	to_scrub.push_back(scrub_job);
-	scrub_job->in_queues = true;
-	scrub_job->state = qu_state_t::registered;
-      }
-      break;
-
-    case qu_state_t::unregistering:
-      // restore to the to_sched queue
-      {
-	// must be under lock, as the job might be removed from the queue
-	// at any minute
-	std::lock_guard lck{jobs_lock};
-
-	update_job(scrub_job, suggested, true /* resets not_before */);
-	if (scrub_job->state == qu_state_t::not_registered) {
-	  dout(5) << " scrub job state changed to 'not registered'" << dendl;
-	  to_scrub.push_back(scrub_job);
-	}
-	scrub_job->in_queues = true;
-	scrub_job->state = qu_state_t::registered;
-      }
-      break;
-  }
-
   dout(10) << fmt::format(
-		"pg[{}] sched-state changed from <{:.14}> to <{:.14}> (@{:s})",
-		scrub_job->pgid, state_at_entry, scrub_job->state.load(),
-		scrub_job->schedule.not_before)
+		  "removing pg[{}] from OSD scrub queue", scrub_job.pgid)
 	   << dendl;
+
+  std::unique_lock lck{jobs_lock};
+  std::erase_if(to_scrub, [&scrub_job](const auto& job) {
+    return job->pgid == scrub_job.pgid;
+  });
 }
 
 
-void ScrubQueue::update_job(Scrub::ScrubJobRef scrub_job,
-			    const sched_params_t& suggested,
-                            bool reset_nb)
+void ScrubQueue::enqueue_target(const Scrub::ScrubJob& sjob)
 {
-  // adjust the suggested scrub time according to OSD-wide status
-  auto adjusted = adjust_target_time(suggested);
-  scrub_job->high_priority = suggested.is_must == must_scrub_t::mandatory;
-  scrub_job->update_schedule(adjusted, reset_nb);
+  if (!sjob.is_registered()) {
+    dout(10) << fmt::format("pg[{}] not scrub-able by this OSD", sjob.pgid) << dendl;
+    return;
+  }
+  ceph_assert(!sjob.target_queued);
```

Comment:
As mentioned above, let's not look at members irrelevant to this component.

Reply:
I will remove. But note: it did serve a purpose in detecting omissions during development.
I'll add the check to all callers.

Reply:
Adding these checks to the point in PgScrubber where it does the scheduling would make sense.

Reply:
After a second thought: I see this differently. The ScrubQueue object maintains the queue, and should
be allowed to access what it queues.
Anyway - will remove it at this phase, as the element queued is changed in the next phase.